README.md
# gssurgo

[![PyPiVersion](https://img.shields.io/pypi/v/gssurgo.svg)](https://pypi.python.org/pypi/gssurgo/) [![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active) [![PYPI Downloads](https://img.shields.io/pypi/dm/gssurgo.svg)](https://pypistats.org/packages/gssurgo) [![Build Status](https://travis-ci.org/jsta/gssurgo.svg?branch=master)](https://travis-ci.org/jsta/gssurgo) [![Code DOI](https://zenodo.org/badge/142450474.svg)](https://zenodo.org/badge/latestdoi/142450474)

The `gssurgo` python package enables open source workflows with the `gSSURGO` dataset. It provides:

* A shell script `extract_gssurgo_tif` for generating stand-alone `gSSURGO` grids. **These raster grids are distributed within file geodatabase archives and can only be extracted using ArcGIS, the fileGDB driver, or (in the case of `extract_gssurgo_tif`) the `arcpy` python package.**  

* Python functions for converting Geodatabase files to geopackage format. 
 
* Python functions for returning the results of specific `SQL` queries of `gSSURGO` data.
 
* Python functions for referencing query results to corresponding (raster) grid cells.

## Prereqs

* The intial `tif` (grid) extraction step requies the `arcpy` python module. This step assumes that a python executable linked to `arcpy` can be found at `C:\Python27\ArcGIS10.3\python.exe`. Edit [bin/extract_gssurgo_tif](bin/extract_gssurgo_tif) to enable alternate locations.

* Remaining operations require the dependencies listed in [environment.yml](environment.yml) and [requirements.txt](requirements.txt). If using Anaconda, make sure you have the **64bit** version. You can install an Anaconda virtual environment with:

```
conda env create -n gssurgo -f environment.yml
source activate gssurgo
```

## Installation

```
# local install
# pip install -e  . 

# development install 
pip install git+git://github.com/jsta/gssurgo.git

# development upgrade
# pip install --upgrade git+git://github.com/jsta/gssurgo.git
```

## Usage

A demonstration workflow using the `gssurgo` python package can be found at: https://github.com/jsta/gssurgo_data

### 1. Extract tif and build gpkgs

```
extract_gssurgo_tif 'path/to/gSSURGO_STATE.gdb/MapunitRaster_10m' 'path/to/STATE.tif'
```

```py
import gssurgo
gssurgo.build_gpkg("path/to/gSSURGO_STATE.gdb", "path/to/gSSURGO_STATE.gpkg")
```

### 2. Generate an Area of Interest (AOI)

```py
gssurgo.aoi(in_raster_path = "tifs", out_raster = "path/to/aoi.tif", xmax = -88.34945, xmin = -88.35470, ymin = 38.70095, ymax = 38.70498)
```

### 3. Pull specific variable and merge with corresponding tif

```py
gssurgo.query_gpkg(src_tif = "tests/aoi.tif", gpkg_path = "path/to/gkpgs/", sql_query = 'SELECT mukey, nonirryield_r FROM mucropyld WHERE (cropname = "Corn")', out_raster = "tests/aoi_results.tif")

gssurgo.query_gpkg(src_tif = "tests/aoi.tif", gpkg_path = "path/to/gpkgs/", sql_query = 'SELECT mukey, nonirryield_r FROM mucropyld WHERE (cropname = "Corn")', out_raster = "tests/aoi_results.tif")

```

> The `sql_query` parameter must give a two column result of `mukey` and `some_variable` where no `mukey` entries are duplicated.

### 4. Visualize output

```py
gssurgo.viz_numeric_output("path/to/aoi_results.tif", "path/to/aoi_results.png")
```

![](scratch/nonirryield_r.png)


CHANGELOG
1.0.1:
    - Name fix

1.0.0:
    - Handle type conversion error #33

gssurgo/aoi.py
"""Functions for defining an area of interest (aoi)."""

import geopandas as gpd
from matplotlib import pyplot as plt
import re
import gdal
import os
import rasterio.merge
from pyproj import Proj, transform
import pkg_resources


def state_by_bbox(fpath, ext, xmax, xmin, ymin, ymax):
    r"""Find state abbrevations from a long-lat bbox.

    Examples
    --------
    gssurgo.state_by_bbox(fpath = "tifs", ext = "tif",  xmax = -88.34945, \
                          xmin = -88.35470, ymin = 38.70095, ymax = 38.70498)

    Notes
    -----
    download:
    http://www2.census.gov/geo/tiger/GENZ2017/shp/cb_2017_us_state_500k.zip
    `ogr2ogr -f GPKG states.gpkg cb_2017_us_state_50k.shp`
    """
    states = gpd.read_file(pkg_resources.resource_filename('gssurgo',
                                                           'states.gpkg'))
    states = states[states.STATEFP.astype(int) < 60]
    states = states[~states.NAME.isin(['Hawaii', 'Alaska'])]

    states = states.cx[xmin:xmax, ymin:ymax]

    # ax = states.plot(color='white', edgecolor='black')
    # plt.show()

    keyword_base = r'.*(%s).' + ext
    keywords = re.compile(keyword_base % '|'.join(states.STUSPS))
    candidate_files = [os.path.join(fpath, f) for
                       f in os.listdir(fpath) if f.endswith('.' + ext)]

    return list(filter(keywords.match, candidate_files))


def aoi(in_raster_path, out_raster, xmin, ymax, xmax, ymin, src_tif = None):
    r"""Crop original raster to bounding box so we can read in memory.

    Examples
    --------
    gssurgo.aoi(in_raster_path = "tifs", out_raster = "path/to/aoi.tif", \
                xmax = -88.34945, xmin = -88.35470, ymin = 38.70095, \
                ymax = 38.70498)

    Notes
    -----
    https://gis.stackexchange.com/a/237412/32531

    """
    if src_tif is None:
        src_tif = state_by_bbox(fpath = in_raster_path, xmax = xmax,
                                xmin = xmin, ymin = ymin, ymax = ymax,
                                ext = "tif")

    # https://gis.stackexchange.com/a/78944/32531
    inProj = Proj(init='epsg:4326')
    outProj = Proj('+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 \
                    +x_0=0 +y_0=0 +ellps=GRS80 +towgs84=0,0,0,0,0,0,0 \
                    +units=m +no_defs')
    xmin, ymin = transform(inProj, outProj, xmin, ymin)
    xmax, ymax = transform(inProj, outProj, xmax, ymax)

    if(len(src_tif) == 1):
        src_tif = ''.join(src_tif)
        ds = gdal.Open(src_tif)
        # raster_res = ds.GetGeoTransform()[1]
        ds = gdal.Translate(out_raster, ds, projWin = [xmin, ymax, xmax, ymin])
    else:
        ds = gdal.Open(''.join(src_tif[0]))
        ds = gdal.Translate("temp1.tif", ds,
                            projWin = [xmin, ymax, xmax, ymin])

        ds = gdal.Open(''.join(src_tif[1]))
        ds = gdal.Translate("temp2.tif", ds,
                            projWin = [xmin, ymax, xmax, ymin])

        # https://gist.github.com/nishadhka/9bc758129c2949a3194b79570198f544
        d1 = rasterio.open("temp1.tif", dtype = "uint32")
        d2 = rasterio.open("temp2.tif", dtype = "uint32")

        dest, output_transform = rasterio.merge.merge([d1, d2])

        with d1 as src:
            out_meta = src.meta.copy()
        out_meta.update({"driver": "GTiff",
                         "height": dest.shape[1],
                         "width": dest.shape[2],
                         "transform": output_transform})
        with rasterio.open(out_raster, "w", **out_meta) as dest1:
            dest1.write(dest)
